src/wccp/WccpConfig.cc

```diff
@@ -367,342 +288,339 @@ operator==(const char *text, ts::ConstBuffer const &b)
+      return ts::Errata::Message(3, LVL_INFO, "No 'wccp' node found in configuration.");
+    }
+
+    return loader(wccp_node);
+  } catch (std::exception &ex) {
+    return ts::Errata::Message(1, 1, ex.what());
+  }
+
+  return ts::Errata();
+}
+
+ts::Errata
+CacheImpl::loader(const YAML::Node &cfg)
 {
   ts::Errata zret;
-  int src_line = 0;              // scratch for local source line caching.
-  std::vector<uint32_t> routers; // scratch per service loop.
-  Value prop;                    // scratch var.
 
-  ts::Rv<Configuration> cv = Configuration::loadFromPath(path);
-  if (!cv.isOK())
-    return cv.errata();
+  YAML::Node svc_list = cfg[SVCS_NAME];
 
-  ts::config::Configuration cfg = cv.result();
-  Value svc_list                = cfg.find("services");
   // No point in going on from here.
-  if (!svc_list)
+  if (!svc_list) {
     return Services_Not_Found();
+  }
 
-  if (!svc_list.isContainer())
+  if (!svc_list.IsSequence()) {
     return Services_Not_A_Sequence();
+  }
 
   // Check for global (default) security setting.
-  if ((prop = cfg[SVC_PROP_SECURITY]).hasValue()) {
-    ts::Rv<ts::ConstBuffer> rv = load_security(prop);
-    if (rv.isOK())
-      this->useMD5Security(rv);
-    else
+  YAML::Node prop_sec = cfg[SVC_PROP_SECURITY];
+  if (prop_sec) {
+    ts::Rv<std::string> rv = load_security(prop_sec);
+    if (rv.isOK() && rv.result().size() > 0) {
+      std::string md5_key = rv.result();
+      this->useMD5Security(ts::ConstBuffer(md5_key.c_str(), md5_key.size()));
+    } else {
       zret.pull(rv.errata());
+    }
   }
 
-  if ((prop = cfg[SVC_PROP_ROUTERS]).hasValue()) {
-    zret.pull(load_routers(prop, Seed_Router).doNotLog());
+  auto prop_routers = cfg[SVC_PROP_ROUTERS];
+  if (prop_routers) {
+    zret.pull(load_routers(prop_routers, Seed_Router).doNotLog());
   }
 
-  int idx, nsvc;
-  for (idx = 0, nsvc = svc_list.childCount(); idx < nsvc; ++idx) {
-    int x; // scratch int.
-    const char *md5_key = 0;
-    ts::ConstBuffer text;
+  for (auto it = svc_list.begin(); it != svc_list.end(); ++it) {
+    std::string md5_key;
     SecurityOption security_style = SECURITY_NONE;
     bool use_group_local_security = false;
-    Value const &svc_cfg          = svc_list[idx];
-    int svc_line                  = svc_cfg.getSourceLine();
-    ServiceGroup svc_info;
 
-    if (ts::config::GroupValue != svc_cfg.getType()) {
-      zret.push(Service_Not_A_Group(svc_line));
-      continue;
-    }
+    const YAML::Node &svc_cfg = *it;
+    ServiceGroup svc_info;
 
     // Get the service ID.
-    if ((prop = svc_cfg[SVC_PROP_ID]).hasValue()) {
-      if (ts::config::IntegerValue == prop.getType()) {
-        x = atoi(prop.getText()._ptr);
-        if (0 <= x && x <= 255)
-          svc_info.setSvcId(x);
-        else
-          zret.push(Svc_Prop_Out_Of_Range(SVC_PROP_ID, prop, x, 0, 255));
+    YAML::Node propId = svc_cfg[SVC_PROP_ID];
+    if (propId) {
+      int x = propId.as<int>();
+      if (0 <= x && x <= 255) {
```

Comment:
What value does `as<int>("ganchovy")` yield? 0? In which case any random string will appear to be valid.

Reply:
An exception is thrown. error at line X, column X: bad conversion